Update switchbot to be local push

bdraco opened the pull request:
## Breaking change

The bluetooth integration must be enabled before switchbot can be used.  Bluetooth is automatically discovered in most cases, but can be manually added from integrations page if it is not.

The `Time between updates (seconds)` and `How long to scan for advertisement data` options have been removed as they are no longer needed since the integration is now push instead of polling.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Update switchbot to be local push.  This prepares the integration to be able to support more devices including the contact sensors, motion sensors, etc.  In the short term, it means curtains and sensors get near instant updates.

Changelog: https://github.com/Danielhiversen/pySwitchbot/compare/0.14.1...0.15.0

Fixes the assumption that address are mac addresses (they are UUIDs on mac os)
<img width="583" alt="Screen Shot 2022-07-23 at 02 10 23" src="https://user-images.githubusercontent.com/663432/180594695-3a7892ec-7139-42cd-8ef1-678b62b2c2c8.png">

There is a problem discovering the temp sensors via bluetooth discovery since the first update doesn't have all the data so it gets rejected, but thats a problem with the bluetooth integration not sending the discovery again since it only does it when the mfr data appears, not the service data.

Setting them up manually via the user flow works just fine though.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/23495

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

bdraco committed:
Update switchbot to be local push

bdraco committed:
fixes

bdraco committed:
fixes

bdraco committed:
fixes

bdraco committed:
fixes

bdraco committed:
adjust

bdraco committed:
cover is not assumed anymore

bdraco committed:
cleanups

bdraco committed:
adjust

bdraco committed:
adjust

bdraco committed:
add missing cover

bdraco committed:
import compat

bdraco committed:
fixes

bdraco committed:
uses lower

bdraco committed:
uses lower

bdraco committed:
bleak users upper case addresses

bdraco committed:
fixes

bdraco committed:
bump

bdraco committed:
Merge branch 'dev' into switchbot_local_push

bdraco commented:
Verified the migration of the mac to address works and all the entries come up on the new verison.

MartinHjelmare commented on homeassistant/components/switchbot/__init__.py:
```diff
@@ -13,73 +25,88 @@
+        hass.config_entries.async_update_entry(
+            entry,
+            data={k: v for k, v in entry.data.items() if k != CONF_MAC}
+            | {CONF_ADDRESS: mac},
```
It's a bit hard to read this when it's split on two lines. Maybe use a parenthesis or local variable for the new data before the call?

bdraco replied:
5a2ed118342b0c49dcc256ee67b6b7ff9caa17c4

MartinHjelmare commented on homeassistant/components/switchbot/__init__.py:
```diff
@@ -13,73 +25,88 @@
+            mac = dr.format_mac(mac)
+        hass.config_entries.async_update_entry(
+            entry,
+            data={k: v for k, v in entry.data.items() if k != CONF_MAC}
```
If possible, we should keep the legacy data in the config entry data to allow roll backs. We can just add new data items and keep the old.

bdraco replied:
5a2ed118342b0c49dcc256ee67b6b7ff9caa17c4

MartinHjelmare commented on homeassistant/components/switchbot/__init__.py:
```diff
@@ -13,73 +25,88 @@
+            options={
+                k: v
+                for k, v in entry.options.items()
+                if k not in DEPRECATED_CONFIG_OPTIONS
```
Try to keep the deprecated data.

bdraco replied:
5a2ed118342b0c49dcc256ee67b6b7ff9caa17c4

MartinHjelmare commented on homeassistant/components/switchbot/entity.py:
```diff
@@ -4,43 +4,70 @@
         self._attr_name = name
         self._attr_device_info = DeviceInfo(
-            connections={(dr.CONNECTION_NETWORK_MAC, self._mac)},
+            identifiers={(bluetooth.DOMAIN, self._address)},
```
Normally we don't use another domain in identifiers.

bdraco replied:
Maybe we should add a CONNECTION_BLUETOOTH to the device registry so integrations that use the same device could link them?

bdraco replied:
https://github.com/home-assistant/core/pull/75666

I'll go back and refactor the coordinator in a followup as well to use it

MartinHjelmare replied:
Yeah, adding a Bluetooth connection sounds good.

bdraco replied:
Adjusted

MartinHjelmare commented on homeassistant/components/switchbot/coordinator.py:
```diff
@@ -22,40 +26,94 @@ def flatten_sensors_data(sensor):
 
 
-class SwitchbotDataUpdateCoordinator(DataUpdateCoordinator):
+class SwitchbotCoordinator:
```
If we split the entity details logic out from the passive bluetooth coordinator it looks like we can re-use much of that here. Inherit from it here and implement overrides as needed.

bdraco replied:
Will refactor after https://github.com/home-assistant/core/pull/75642 merges

bdraco replied:
I think we might need a more generic processor class for this case but I can refactor the existing one to subclass a newer more generic one

bdraco replied:
I think we can separate the processor into PassiveBluetoothDataProcessor and PassiveBluetoothEntityDataProcessor

bdraco replied:
Not so much reusable but 3 of the functions are so I'll make it a base class instead and then subclass it here to use it as a processor for switch bot

MartinHjelmare replied:
I was just thinking we can reuse the coordinator, not the processor.

bdraco replied:
I refactored so there are two coordinator options

`PassiveBluetoothDataUpdateCoordinator`, similar to the existing `DataUpdateCoordinator` but the caller needs to subclass `_async_handle_bluetooth_event` to handle the data

`PassiveBluetoothProcessorCoordinator`, what used to be called `PassiveBluetoothDataUpdateCoordinator` and is used for the `inkbird`/`xiaomi_ble`/`sensorpush` so far.

bdraco replied:
I was thinking of moving all the processor based ones into `passive_update_processor.py`

MartinHjelmare replied:
Looks good. I don't have an opinion about moving things.

bdraco replied:
I'll move it around so I don't confused the two of them and then make a new PR

bdraco committed:
keep conf_mac and deprecated options for rollback

bdraco committed:
Merge branch 'dev' into switchbot_local_push

bdraco committed:
reuse coordinator

bdraco committed:
adjust

bdraco committed:
move around

bdraco committed:
move around

bdraco committed:
move around

bdraco committed:
move around

MartinHjelmare commented on homeassistant/components/bluetooth/passive_update_processor.py:
```diff
@@ -0,0 +1,346 @@
+"""The Bluetooth integration."""
```
```suggestion
"""Passive update processors for the Bluetooth integration."""
```

MartinHjelmare commented on homeassistant/components/bluetooth/passive_update_processor.py:
```diff
@@ -0,0 +1,346 @@
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import BluetoothChange
+from ._coordinator import BasePassiveBluetoothCoordinator
```
```suggestion
from .update_coordinator import BasePassiveBluetoothCoordinator
```

MartinHjelmare commented on homeassistant/components/bluetooth/update_coordinator.py:
```diff
@@ -0,0 +1,84 @@
+"""The Bluetooth integration."""
```
```suggestion
"""Update coordinator for the Bluetooth integration."""
```

MartinHjelmare commented on homeassistant/components/bluetooth/passive_update_coordinator.py:
```diff
@@ -1,130 +1,16 @@
 """The Bluetooth integration."""
```
```suggestion
"""Passive update coordinator the Bluetooth integration."""
```

bdraco committed:
refactor fixes

bdraco committed:
compat with DataUpdateCoordinator

bdraco committed:
fix available

bdraco committed:
Update homeassistant/components/bluetooth/passive_update_processor.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bdraco committed:
Update homeassistant/components/bluetooth/passive_update_coordinator.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bdraco committed:
Update homeassistant/components/bluetooth/update_coordinator.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bdraco commented:
Everything is back to working again.  I'm going to pull this out into a fresh PR and add some more tests

bdraco committed:
Split bluetooth coordinator into PassiveBluetoothDataUpdateCoordinator and PassiveBluetoothProcessorCoordinator

The PassiveBluetoothDataUpdateCoordinator is now used to replace instances
of DataUpdateCoordinator where the data is coming from bluetooth
advertisements, and the integration may also mix in active updates

The PassiveBluetoothProcessorCoordinator is used for integrations that
want to process each bluetooth advertisement with multiple processors
which can be dispatched to individual platforms or areas or the integration
as it chooes

bdraco commented:
I pulled out the coordinator changes into https://github.com/home-assistant/core/pull/75675

Working on some tests now

bdraco committed:
Merge branch 'update_coordinator_split' into switchbot_local_push

bdraco committed:
change connections

bdraco committed:
reduce code churn to reduce review overhead

bdraco committed:
reduce code churn to reduce review overhead

bdraco committed:
Update homeassistant/components/bluetooth/passive_update_coordinator.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bdraco committed:
add basic test

bdraco committed:
add basic test

bdraco committed:
Merge branch 'update_coordinator_split' into switchbot_local_push

bdraco committed:
complete coverage

bdraco committed:
Merge branch 'update_coordinator_split' into switchbot_local_push

bdraco committed:
Merge branch 'dev' into switchbot_local_push

MartinHjelmare commented on homeassistant/components/switchbot/coordinator.py:
```diff
@@ -22,40 +28,53 @@ def flatten_sensors_data(sensor):
+        logger: logging.Logger,
+        ble_device: BLEDevice,
+        device: switchbot.SwitchbotDevice,
+        common_options: dict[str, int],
```
```suggestion
        common_options: Mapping[str, int],
```

MartinHjelmare commented on homeassistant/components/switchbot/coordinator.py:
```diff
@@ -1,15 +1,21 @@
 from __future__ import annotations
 
-from datetime import timedelta
+import asyncio
```
```suggestion
import asyncio
from collections.abc import Mapping
```

MartinHjelmare commented on homeassistant/components/switchbot/__init__.py:
```diff
@@ -29,57 +37,67 @@
 
-    await coordinator.async_config_entry_first_refresh()
+    if COMMON_OPTIONS not in domain_data:
+        domain_data[COMMON_OPTIONS] = entry.options
```
I don't understand why we need to store the config entry options in hass.data. Are these still really common options?

bdraco replied:
I think we can get rid of them but right now we have `switchbot.DEFAULT_RETRY_TIMEOUT = common_options[CONF_RETRY_TIMEOUT]`

I'll do another PR to remove them if https://github.com/Danielhiversen/pySwitchbot/pull/50 merges in time for beta

bdraco replied:
https://github.com/home-assistant/core/pull/75692

MartinHjelmare commented on homeassistant/components/switchbot/__init__.py:
```diff
@@ -29,57 +37,67 @@
+        domain_data[COMMON_OPTIONS] = entry.options
 
-    entry.async_on_unload(entry.add_update_listener(_async_update_listener))
+    common_options: dict[str, int] = domain_data[COMMON_OPTIONS]
```
```suggestion
    common_options: Mapping[str, int] = domain_data[COMMON_OPTIONS]
```

bdraco replied:
There is no point in having common options anymore since each one can have its own.  And that makes more sense since the battery powered devices likely might need different adjustments if they have range issues

bdraco replied:
The retry timeout is likely because the connect timeout is too low which means [services cannot be resolved in time](https://github.com/hbldh/bleak/blob/0361ba4b7e4682f4e97ecacaab4499b84c70f288/bleak/backends/bluezdbus/client.py#L316) so it disconnects in the middle of the connection leaving things in a bad state

bdraco replied:
I opened a pr with the lib to fix that https://github.com/Danielhiversen/pySwitchbot/pull/50

MartinHjelmare commented on homeassistant/components/switchbot/__init__.py:
```diff
@@ -1,10 +1,23 @@
 """Support for Switchbot devices."""
 
+import logging
```
```suggestion
from collections.abc import Mapping
import logging
```

bdraco committed:
Update homeassistant/components/switchbot/coordinator.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bdraco committed:
Update homeassistant/components/switchbot/coordinator.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bdraco committed:
Update homeassistant/components/switchbot/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bdraco committed:
Update homeassistant/components/switchbot/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bdraco committed:
lint

bdraco commented:
Thanks.

bdraco commented:
I'm going to fix the unreliable discovery in bluetooth next so the temp sensors get discovered and the other BLE integration that rely on service data have discovery

bdraco commented:
Test failure is unrelated to this PR
`FAILED tests/components/litejet/test_trigger.py::test_held_more_than_short[pyloop]`
